salt/returners/mongo_future_return.py

```diff
@@ -141,15 +155,22 @@ def _get_conn(ret):
     # at some point we should remove support for
     # pymongo versions < 2.3 until then there are
     # a bunch of these sections that need to be supported
-
-    if float(version) > 2.3:
-        conn = pymongo.MongoClient(host, port)
+    if uri and float(version) > 2.3:
+        if uri and host:
+            warnlog = 'Specified both URI and host. URI {0} will be taken'.format(uri)
+            log.warn(warnlog)
+        pymongo.uri_parser.parse_uri(uri)
+        conn = pymongo.MongoClient(uri)
+        mdb = conn.get_database()
     else:
-        conn = pymongo.Connection(host, port)
-    mdb = conn[db_]
+        if float(version) > 2.3:
+            conn = pymongo.MongoClient(host, port)
+        else:
+            conn = pymongo.Connection(host, port)
 
-    if user and password:
-        mdb.authenticate(user, password)
+        mdb = conn[db_]
+        if user and password:
+            mdb.authenticate(user, password)
 
     if indexes:
         if float(version) > 2.3:
```

Comment:
If you wouldn't mind, it would be great to move all these other version comparisons to using LooseVersion as well.